finality-aleph/src/testing/proposal.rs

```diff
@@ -0,0 +1,270 @@
+        );
+    }
+}
+
+#[tokio::test]
+async fn incorrect_branch_handled_correctly() {
+    let (mut chain_builder, mut cached_cip, mut aux_cip) = prepare_proposal_test();
+
+    let blocks = chain_builder
+        .build_and_import_branch_upon(
+            &chain_builder.genesis_hash(),
+            MAX_DATA_BRANCH_LEN * 10,
+            false,
+        )
+        .await;
+
+    let incorrect_branch = vec![
+        blocks[0].clone(),
+        blocks[1].clone(),
+        blocks[3].clone(),
+        blocks[5].clone(),
+    ];
+    let proposal = proposal_from_blocks(incorrect_branch);
+    verify_proposal_status(
+        &mut cached_cip,
+        &mut aux_cip,
+        &proposal,
+        Pending(TopBlockImportedButIncorrectBranch),
+    );
+
+    chain_builder.finalize_block(&blocks[1].header.hash());
+    verify_proposal_status(
+        &mut cached_cip,
+        &mut aux_cip,
+        &proposal,
+        Pending(TopBlockImportedButIncorrectBranch),
+    );
+
+    chain_builder.finalize_block(&blocks[10].header.hash());
+    verify_proposal_status(&mut cached_cip, &mut aux_cip, &proposal, Ignore);
+}
+
+#[tokio::test]
+async fn pending_top_block_handled_correctly() {
+    let (mut chain_builder, mut cached_cip, mut aux_cip) = prepare_proposal_test();
+
+    let blocks = chain_builder
+        .build_branch_upon(&chain_builder.genesis_hash(), MAX_DATA_BRANCH_LEN * 10)
+        .await;
+
+    for len in 1..=MAX_DATA_BRANCH_LEN {
+        let blocks_branch = blocks[0..len].to_vec();
+        let proposal = proposal_from_blocks(blocks_branch);
+        verify_proposal_status(
+            &mut cached_cip,
+            &mut aux_cip,
+            &proposal,
+            Pending(PendingTopBlock),
+        );
+    }
+    chain_builder.import_branch(blocks.clone(), false).await;
+
+    for len in 1..=MAX_DATA_BRANCH_LEN {
+        let blocks_branch = blocks[0..len].to_vec();
+        let proposal = proposal_from_blocks(blocks_branch);
+        verify_proposal_status(
+            &mut cached_cip,
+            &mut aux_cip,
+            &proposal,
+            Finalize(proposal.top_block()),
+        );
+    }
+}
+
+#[tokio::test]
+async fn hopeless_forks_handled_correctly() {
+    let (mut chain_builder, mut cached_cip, mut aux_cip) = prepare_proposal_test();
+
+    let blocks = chain_builder
+        .build_and_import_branch_upon(
+            &chain_builder.genesis_hash(),
+            MAX_DATA_BRANCH_LEN * 10,
+            false,
+        )
+        .await;
+
+    let forks = chain_builder
+        .build_branch_upon(&blocks[2].header.hash(), MAX_DATA_BRANCH_LEN * 10)
+        .await;
+
+    for len in 1..=MAX_DATA_BRANCH_LEN {
+        let fork_branch = forks[0..len].to_vec();
+        let proposal = proposal_from_blocks(fork_branch);
+        verify_proposal_status(
+            &mut cached_cip,
+            &mut aux_cip,
+            &proposal,
+            Pending(PendingTopBlock),
+        );
+    }
```

Comment:
Fine, extracted a function. But just for the record: I hate such functions :P They have lots of random arguments, confusing semantics and IMHO they make the code harder to read. And the only reason for having them is the fear of code duplication.

Reply:
well, I do not want to introduce more frustration, but since we already made a move in this direction, we could reduce the duplication even more, by changing `verify_proposal_of_all_lens_finalizable` to `verify_proposal_of_all_lens_of_status` and adding one (not so random :stuck_out_tongue:) argument `status` :smiley: